flytekit/core/data_persistence.py

```diff
@@ -0,0 +1,382 @@
+    runs.
+    """
+
+    PROTOCOL = "file://"
+
+    def __init__(self, default_prefix: typing.Optional[str] = None, **kwargs):
+        super().__init__(name="local", default_prefix=default_prefix, **kwargs)
+
+    @staticmethod
+    def _make_local_path(path):
+        if not os.path.exists(path):
+            try:
+                pathlib.Path(path).mkdir(parents=True, exist_ok=True)
+            except OSError:  # Guard against race condition
+                if not os.path.isdir(path):
+                    raise
+
+    @staticmethod
+    def strip_file_header(path: str) -> str:
+        """
+        Drops file:// if it exists from the file
+        """
+        if path.startswith("file://"):
+            return path.replace("file://", "", 1)
+        return path
+
+    def listdir(self, path: str, recursive: bool = False) -> typing.Generator[str, None, None]:
+        if not recursive:
+            files = os.listdir(self.strip_file_header(path))
+            for f in files:
+                yield f
+            return
+
+        for root, subdirs, files in os.walk(self.strip_file_header(path)):
+            for f in files:
+                yield os.path.join(root, f)
+        return
+
+    def exists(self, path: str):
+        return os.path.exists(self.strip_file_header(path))
+
+    def get(self, from_path: str, to_path: str, recursive: bool = False):
+        if from_path != to_path:
+            if recursive:
+                _dir_util.copy_tree(self.strip_file_header(from_path), self.strip_file_header(to_path))
+            else:
+                copyfile(self.strip_file_header(from_path), self.strip_file_header(to_path))
+
+    def put(self, from_path: str, to_path: str, recursive: bool = False):
+        if from_path != to_path:
+            if recursive:
+                _dir_util.copy_tree(self.strip_file_header(from_path), self.strip_file_header(to_path))
+            else:
+                # Emulate s3's flat storage by automatically creating directory path
+                self._make_local_path(os.path.dirname(self.strip_file_header(to_path)))
+                # Write the object to a local file in the sandbox
+                copyfile(self.strip_file_header(from_path), self.strip_file_header(to_path))
+
+    def construct_path(self, _: bool, add_prefix: bool, *args) -> str:
+        # Ignore add_protocol for now. Only complicates things
+        if add_prefix:
+            return os.path.join(self.default_prefix, *args)
+        return os.path.join(*args)
+
+
+class FileAccessProvider(object):
+    """
+    This is the class that is available through the FlyteContext and can be used for persisting data to the remote
+    durable store.
+    """
+
+    def __init__(self, local_sandbox_dir: Union[str, os.PathLike], raw_output_prefix: str):
+        # Local access
+        if local_sandbox_dir is None or local_sandbox_dir == "":
+            raise Exception("Can't use empty path")
```

Comment:
a more specific error?

```suggestion
            raise TypeError("Can't use empty path")
```

Reply:
done